prefetch-yarn-deps: add `fixup-yarn-lock` command to fixup a yarn.lock

lilyinstarlight opened the pull request:
###### Description of changes

I need this because `fixup_yarn_lock` doesn't necessarily parse all yarn.lock files correctly ([some versions of npm will rewrite yarn.lock files to have more quoting](https://github.com/npm/cli/issues/5126)...) and also does not clear out `integrity` for git deps. Clearing out `integrity` is needed for git deps since the hash recorded in the lockfile will not match the prefetched deps that are repacked to be more reproducible (this matches the behavior for `prefetch-npm-deps --fixup-lockfile`)

The code was simple enough and `prefetch-yarn-deps` was already pulling in Yarn's actual lockfile parsing library (~~and also `prefetch-yarn-deps` now mirrors `prefetch-npm-deps --fixup-lockfile`~~), so I added it to the existing `prefetch-yarn-deps` package instead of rewriting `fixup_yarn_lock` from yarn2nix

cc: @yu-re-ka @winterqt 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

###### nixpkgs-review

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>prefetch-yarn-deps</li>
  </ul>
</details>

yu-re-ka commented:
I would prefer to keep the fixup lockfile separate to keep the part that runs inside the Fixed-Output Derivations as simple as possible

lilyinstarlight commented:
> I would prefer to keep the fixup lockfile separate to keep the part that runs inside the Fixed-Output Derivations as simple as possible. Maybe it can reference shared code though.

That code doesn't run in a FOD, but I'll factor it out regardless some time in the next few days when I get a chance

lilyinstarlight commented:
The only common code really is `urlToName`, so I could factor that out into a separate file in addition to the fixup code and create two binaries in the `prefetch-yarn-deps` derivation (this would duplicate entrypoint code between `index.js` and `fixup.js`, but that's not too bad I guess)

Is that something like what you were thinking @yu-re-ka?

Thanks for the feedback!

<details>
<summary>potential-refactor.patch</summary>

```diff
diff --git a/pkgs/build-support/node/fetch-yarn-deps/default.nix b/pkgs/build-support/node/fetch-yarn-deps/default.nix
index 4cf2507706a..f3b5a6bd163 100644
--- a/pkgs/build-support/node/fetch-yarn-deps/default.nix
+++ b/pkgs/build-support/node/fetch-yarn-deps/default.nix
@@ -21,8 +21,8 @@ in {
       mkdir libexec
       tar --strip-components=1 -xf ${yarnpkg-lockfile-tar} package/index.js
       mv index.js libexec/yarnpkg-lockfile.js
-      cp ${./index.js} libexec/index.js
-      patchShebangs libexec/index.js
+      cp ${./.}/*.js libexec/
+      patchShebangs libexec
 
       runHook postBuild
     '';
@@ -34,6 +34,7 @@ in {
       cp -r libexec $out
       makeWrapper $out/libexec/index.js $out/bin/prefetch-yarn-deps \
         --prefix PATH : ${lib.makeBinPath [ coreutils nix-prefetch-git nix ]}
+      makeWrapper $out/libexec/fixup.js $out/bin/fixup-yarn-lock
 
       runHook postInstall
     '';
diff --git a/pkgs/build-support/node/fetch-yarn-deps/fixup.js b/pkgs/build-support/node/fetch-yarn-deps/fixup.js
new file mode 100755
index 00000000000..6f22d8fb2dc
--- /dev/null
+++ b/pkgs/build-support/node/fetch-yarn-deps/fixup.js
@@ -0,0 +1,72 @@
+#!/usr/bin/env node
+'use strict'
+
+const fs = require('fs')
+const process = require('process')
+const lockfile = require('./yarnpkg-lockfile.js')
+const urlToName = require('./urlToName.js')
+
+const fixupYarnLock = async (lockContents, verbose) => {
+    const lockData = lockfile.parse(lockContents)
+
+    const fixedData = Object.fromEntries(
+        Object.entries(lockData.object)
+        .map(([dep, pkg]) => {
+            const [ url, sha1 ] = pkg.resolved.split("#", 2)
+
+            if (sha1 || url.startsWith("https://codeload.github.com")) {
+                if (verbose) console.log(`Removing integrity for git dependency ${dep}`)
+                delete pkg.integrity
+            }
+
+            if (verbose) console.log(`Rewriting URL ${url} for dependency ${dep}`)
+            pkg.resolved = urlToName(url)
+
+            return [dep, pkg]
+        })
+    )
+
+    await fs.promises.writeFile('yarn.lock', lockfile.stringify(fixedData))
+    if (verbose) console.log('Done')
+}
+
+const showUsage = async () => {
+	process.stderr.write(`
+syntax: fixup-yarn-lock [path to yarn.lock] [options]
+
+Options:
+  -h --help         Show this help
+  -v --verbose      Verbose output
+`)
+	process.exit(1)
+}
+
+const main = async () => {
+	const args = process.argv.slice(2)
+	let next, lockFile, verbose
+	while (next = args.shift()) {
+		if (next == '--verbose' || next == '-v') {
+			verbose = true
+		} else if (next == '--help' || next == '-h') {
+			showUsage()
+		} else if (!lockFile) {
+			lockFile = next
+		} else {
+			showUsage()
+		}
+	}
+	let lockContents
+	try {
+		lockContents = await fs.promises.readFile(lockFile || 'yarn.lock', 'utf-8')
+	} catch {
+		showUsage()
+	}
+
+	await fixupYarnLock(lockContents, verbose)
+}
+
+main()
+	.catch(e => {
+		console.error(e)
+		process.exit(1)
+	})
diff --git a/pkgs/build-support/node/fetch-yarn-deps/index.js b/pkgs/build-support/node/fetch-yarn-deps/index.js
index 93f897a3a1b..4abc79b1e29 100755
--- a/pkgs/build-support/node/fetch-yarn-deps/index.js
+++ b/pkgs/build-support/node/fetch-yarn-deps/index.js
@@ -10,6 +10,7 @@ const path = require('path')
 const lockfile = require('./yarnpkg-lockfile.js')
 const { promisify } = require('util')
 const url = require('url')
+const urlToName = require('./urlToName.js')
 
 const execFile = promisify(child_process.execFile)
 
@@ -19,20 +20,6 @@ const exec = async (...args) => {
 	return res
 }
 
-// This has to match the logic in pkgs/development/tools/yarn2nix-moretea/yarn2nix/lib/urlToName.js
-// so that fixup_yarn_lock produces the same paths
-const urlToName = url => {
-	const isCodeloadGitTarballUrl = url.startsWith('https://codeload.github.com/') && url.includes('/tar.gz/')
-
-	if (url.startsWith('git+') || isCodeloadGitTarballUrl) {
-		return path.basename(url)
-	} else {
-		return url
-			.replace(/https:\/\/(.)*(.com)\//g, '') // prevents having long directory names
-			.replace(/[@/%:-]/g, '_') // replace @ and : and - and % characters with underscore
-	}
-}
-
 const downloadFileHttps = (fileName, url, expectedHash, hashType = 'sha1') => {
 	return new Promise((resolve, reject) => {
 		https.get(url, (res) => {
@@ -144,30 +131,6 @@ const prefetchYarnDeps = async (lockContents, verbose) => {
 	if (verbose) console.log('Done')
 }
 
-const fixupYarnLock = async (lockContents, verbose) => {
-    const lockData = lockfile.parse(lockContents)
-
-    const fixedData = Object.fromEntries(
-        Object.entries(lockData.object)
-        .map(([dep, pkg]) => {
-            const [ url, sha1 ] = pkg.resolved.split("#", 2)
-
-            if (sha1 || url.startsWith("https://codeload.github.com")) {
-                if (verbose) console.log(`Removing integrity for git dependency ${dep}`)
-                delete pkg.integrity
-            }
-
-            if (verbose) console.log(`Rewriting URL ${url} for dependency ${dep}`)
-            pkg.resolved = urlToName(url)
-
-            return [dep, pkg]
-        })
-    )
-
-    await fs.promises.writeFile('yarn.lock', lockfile.stringify(fixedData))
-    if (verbose) console.log('Done')
-}
-
 const showUsage = async () => {
 	process.stderr.write(`
 syntax: prefetch-yarn-deps [path to yarn.lock] [options]
@@ -176,19 +139,16 @@ Options:
   -h --help         Show this help
   -v --verbose      Verbose output
   --builder         Only perform the download to current directory, then exit
-  --fixup-lockfile  Perform fixup on lockfile, then exit
 `)
 	process.exit(1)
 }
 
 const main = async () => {
 	const args = process.argv.slice(2)
-	let next, lockFile, verbose, isBuilder, fixupLockfile
+	let next, lockFile, verbose, isBuilder
 	while (next = args.shift()) {
 		if (next == '--builder') {
 			isBuilder = true
-		} else if (next == '--fixup-lockfile') {
-			fixupLockfile = true
 		} else if (next == '--verbose' || next == '-v') {
 			verbose = true
 		} else if (next == '--help' || next == '-h') {
@@ -206,9 +166,7 @@ const main = async () => {
 		showUsage()
 	}
 
-	if (fixupLockfile) {
-		await fixupYarnLock(lockContents, verbose)
-	} else if (isBuilder) {
+	if (isBuilder) {
 		await prefetchYarnDeps(lockContents, verbose)
 	} else {
 		const { stdout: tmpDir } = await exec('mktemp', [ '-d' ])
diff --git a/pkgs/build-support/node/fetch-yarn-deps/urlToName.js b/pkgs/build-support/node/fetch-yarn-deps/urlToName.js
new file mode 100644
index 00000000000..8238b15530f
--- /dev/null
+++ b/pkgs/build-support/node/fetch-yarn-deps/urlToName.js
@@ -0,0 +1,17 @@
+const path = require('path')
+
+// This has to match the logic in pkgs/development/tools/yarn2nix-moretea/yarn2nix/lib/urlToName.js
+// so that fixup_yarn_lock produces the same paths
+const urlToName = url => {
+	const isCodeloadGitTarballUrl = url.startsWith('https://codeload.github.com/') && url.includes('/tar.gz/')
+
+	if (url.startsWith('git+') || isCodeloadGitTarballUrl) {
+		return path.basename(url)
+	} else {
+		return url
+			.replace(/https:\/\/(.)*(.com)\//g, '') // prevents having long directory names
+			.replace(/[@/%:-]/g, '_') // replace @ and : and - and % characters with underscore
+	}
+}
+
+module.exports = urlToName;
```
</details>

lilyinstarlight commented:
That last push contained a minor change that made sure the lockfile being fixed up gets written back to that same lockfile (rather than just `yarn.lock`)

yu-re-ka commented:
> gentle ping if you have some time to comment on the above refactor to see if that was more of what you were wanting

Yes, please apply the refactor patch

yu-re-ka commented:
I would name the file that contains urlToName function `common.js` so that it is obvious it is used by both scripts

lilyinstarlight commented:
> I would name the file that contains urlToName function `common.js` so that it is obvious it is used by both scripts

I had originally named it that but changed it when it turned out the only common code was a single function and so I made it a module instead

I'll change the export and stuff to rename back to `common.js` so it's not just a function export anymore

lilyinstarlight committed:
prefetch-yarn-deps: add --fixup-lockfile flag to fixup a yarn.lock

The flag iterates through the lockfile entries, rewrites `resolved` URLs
to those that will be in the cache (like `fixup_yarn_lock` from
yarn2nix), removes `integrity` for git deps whose hash won't match the
reproducible repacking that the fetcher does, writes the amended
lockfile, and exits.

lilyinstarlight commented:
> Yes, please apply the refactor patch

Done!

lilyinstarlight commented:
@yu-re-ka, gentle ping again if you have time to review. Thank you!

This could really help stuff like #210814, and also I plan to try to wrangle yarn berry tooling soon which would be based off of this (since as far as I'm concerned, yarn2nix is not worth updating for berry)

yu-re-ka commented:
@ofborg build gitlab.assets schildichat-desktop akkoma-fe

yu-re-ka reviewed (approved):
the new code looks sane

lilyinstarlight commented:
> Do we want to _replace_ the fixup_yarn_lock with this new version, or have both in parallel? Should existing users of fixup_yarn_lock be moved over?

Ideally all users should be able to move, but given it works differently to `fixup_yarn_lock`, it could cause problems. Also `fixup_yarn_lock` is part of yarn2nix so until we get rid of yarn2nix it'll still be around (but we can maybe do a tree-wide change and remove the top-level attr for it)

I'll think about it and open a tracking issue for doing that at some point (I really just need to make a node project board tbh)

yu-re-ka commented:
`gitlab.assets`, `schildichat-web`, `element-desktop.seshat` and `akkoma-fe` are some examples that _don't_ use any component of yarn2nix other than fixup_yarn_lock, so they could be a good place to start the migration
